design/authorization-handler.md

```diff
@@ -0,0 +1,601 @@
+
+At the initial step, the authorization data sent by the client can be retrieved from the initial Authorization-Endpoint request. This retrieved data will be referred to as AuthorizationRequest, which means the initial request to Keycloak to start the authentication process.
+
+SAML, on the other hand, would still only the support the default client scopes as it does today in Keycloak 15 or earlier.
+
+Additionally to everything said above, the `AuthorizationDetails` object will be validated at the beginning of the request, just like Scopes are validated currently in Keycloak by `AuthorizationEndpointChecker.checkValidScope()`.
+
+### How to parse authorization data
+
+The authorization data is available in the OIDC request by 2 ways:
+
+* In the scope parameter. The scope parameter is already supported in current Keycloak, but rather in static way
+
+* In the authorization_details parameter as defined in the [RAR specification](https://datatracker.ietf.org/doc/html/draft-ietf-oauth-rar-07).
+
+Once Keycloak parses the data from any of these 2 parameters, it will save them in the ClientAuthorizationContext object. Note that same request can contain both `scope` and 
+`authorization_details` parameter and hence Keycloak should be able to parse both of them in the same request. The idea is to unify the data in the ClientAuthorizationContext under some structure, which will serve as an abstraction layer over the data that was sent via the scope parameter or authorization_details. However Keycloak will still need to have some flag in ClientAuthorizationContext to track how the data was sent. This may be needed for various reasons. 
+
+For example [Grant Management API specification - section 6.4](https://openid.net/specs/fapi-grant-management-01.html#section-6.4) (not yet supported by Keycloak) allows clients to query Keycloak about the granted consents. Note that the returned structure contains both scope and authorization_details and hence it is needed by KC to track how the particular request was sent.
+
+### ClientAuthorizationContext object
+
+Output of the parsing is ClientAuthorizationContext object, which is a list of AuthorizationDetailsEntry objects. Each AuthorizationDetailsEntry contains data like:
+
+* JSON with the authorization details
+* ClientAuthorizationHandler used
+* Source - the parsing method (scope parameter or authorization_details parameter)
+
+#### Java pseudo-structure
+
+```java
+class ClientAuthorizationContext {
+	/**
+ 	*  List of entries of parsed values from "scope" parameter and    
+      * from "authorization_details" .
+ 	*  Each entry represents single parsed value from "scope"
+      * parameter or single entry of "authorization_details"
+ 	*
+ 	* There are also entries corresponding to all "default client
+     * scopes" for the particular client, which are added
+     * automatically.
+ 	*/
+	private List<AuthorizationDetailsEntry> entries;
+}
+
+class AuthorizationDetailsEntry {
```

Comment:
Wouldn't be better in terms of design if we introduce an interface and specific sub-types for scope-based access and json-based access formats?